### PR TITLE
Make use of `dskit/cache` package

### PR DESCRIPTION
**Pull request by ortuman**
#### What this PR does

Replace `cache` and `cacheutils` packages by `dskit/cache` package. 

#### Which issue(s) this PR fixes or relates to

Fixes N/A

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: vendor/github.com/grafana/dskit/cache/lru.go [hunk 1]
```diff
@@ -49,19 +47,19 @@ func WrapWithLRUCache(c Cache, name string, reg prometheus.Registerer, lruSize i
 		defaultTTL: defaultTTL,
 
 		requests: promauto.With(reg).NewCounter(prometheus.CounterOpts{
-			Name:        "cortex_cache_memory_requests_total",
+			Name:        "cache_memory_requests_total",
```
Comment (pracucci) [inline vendor/github.com/grafana/dskit/cache/lru.go:50]: This is a breaking change for Mimir. It's fine doing this change in dskit, but then in Mimir you should wrap the registerer to add the `cortex_` prefix back.

File: vendor/github.com/grafana/dskit/cache/memcached.go [hunk 2]
```diff
@@ -29,21 +22,21 @@ type MemcachedCache struct {
...
-func NewMemcachedCache(name string, logger log.Logger, memcached cacheutil.RemoteCacheClient, reg prometheus.Registerer) *MemcachedCache {
+func NewMemcachedCache(name string, logger log.Logger, memcached RemoteCacheClient, reg prometheus.Registerer) *MemcachedCache {
 	c := &MemcachedCache{
 		logger:    logger,
 		memcached: memcached,
 		name:      name,
 	}
 
 	c.requests = promauto.With(reg).NewCounter(prometheus.CounterOpts{
-		Name:        "thanos_cache_memcached_requests_total",
```
Comment (pracucci) [inline vendor/github.com/grafana/dskit/cache/memcached.go:40]: This is a breaking change for Mimir. It's fine doing this change in dskit, but then in Mimir you should wrap the registerer to add the `thanos_` prefix back.

File: vendor/github.com/grafana/dskit/cache/memcached_client.go [hunk 3]
```diff
@@ -264,7 +259,7 @@ func newMemcachedClient(
 ) (*memcachedClient, error) {
 	addressProvider := dns.NewProvider(
 		logger,
-		prometheus.WrapRegistererWithPrefix("thanos_memcached_", reg),
+		prometheus.WrapRegistererWithPrefix("memcached_", reg),
```
Comment (pracucci) [inline vendor/github.com/grafana/dskit/cache/memcached_client.go:262]: This is a breaking change for Mimir. It's fine doing this change in dskit, but then in Mimir you should wrap the registerer to add the `thanos_` prefix back.

File: vendor/github.com/grafana/dskit/gate/gate.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
+package gate
```
Comment (pracucci) [inline vendor/github.com/grafana/dskit/gate/gate.go:1]: The code in this file should be removed from Mimir and we should use the one from dskit instead.

**Review by pracucci — COMMENTED** [review]
Thanks Miguel for working on this. I left few comments. I would like to understand why we have to update so many dependencies (e.g. alertmanager?) and what's actually strictly required. For whatever library we have to update, I would like you to check the release notes / changelogs to see what has changed and if anything could impact us.

File: pkg/storage/tsdb/caching_config.go [hunk 5]
```diff
@@ -105,7 +106,7 @@ func CreateCachingBucket(chunksConfig ChunksCacheConfig, metadataConfig Metadata
 	}
```
Comment (pracucci) [inline pkg/storage/tsdb/caching_config.go:106]: Above, I think you need to add the `thanos_` prefix to the registry passed to `cache.CreateClient("chunks-cache")` and `cache.CreateClient("metadata-cache")`. If I'm right, then I'm suprised no test failed. Is there any existing test where we can improve assertions?

Pay attention that below `reg` is also used, so just replacing `reg` with a wrapped one is not correct, but we just need to wrap the `reg` passed to `cache.CreateClient("chunks-cache")` and `cache.CreateClient("metadata-cache")`.

**Review by pracucci — COMMENTED** [review]
I checked all changes in `vendor/` and everything looks safe to me.

File: pkg/storage/tsdb/caching_config.go (hunk 5 above)
Comment (ortuman) [reply pkg/storage/tsdb/caching_config.go:106]: Added missing `thanos_` metric prefix for chunks and metadata caches, and improved integration test assertions. 

done @ 3226c2c9c949ac10d764d0f93b27788c932d7f69

**Pull request merged**